### ISSUE TITLE
PCH Errors

### ISSUE THREAD

**Issue by ax3l** [ComputationalRadiationPhysics/alpaka#383]
I am experimenting with pre-compiled headers (PCH) with [cotire](https://github.com/sakra/cotire) and get the following errors from alpaka (latest version in `develop`: 90ae873b9984071e64343995e7721bde6a13ba7f)

```
alpaka/include/alpaka/atomic/AtomicOmpCritSec.hpp:129:7: warning: extra tokens at end of #endif directive [-Wendif-labels]
In file included from alpaka/include/alpaka/alpaka.hpp:81:0,
alpaka/include/alpaka/block/sync/BlockSyncBarrierOmp.hpp:192:7: warning: extra tokens at end of #endif directive [-Wendif-labels]
                 }
       ^
alpaka/include/alpaka/alpaka.hpp:132:0,
alpaka/include/alpaka/exec/ExecCpuOmp2Blocks.hpp:172:41: internal compiler error: unspellable token PRAGMA_EOL
```

Any idea what could cause them? Confusion on OpenMP pragmas?

**Comment by BenjaminW3** [comment]
This really seems to be caused by OpenMP. However, I have looked at the given lines and there is nothing strange.
Some ideas:
* Could you try to add some additional newlines at the end of the given files?
* Could you verify that the line-endings in the headers are all unix line endings?
* Could you save and look at the full preprocessed PCH file?

**Comment by ax3l** [comment]
Thanks!

> Could you try to add some additional newlines at the end of the given files?

hm, no change.
But admittedly, the "warnings" are triggered by the last `}` and not the `#endif` line

> Could you verify that the line-endings in the headers are all unix line endings?

`dos2unix` does not change the file, should be fine.

> Could you save and look at the full preprocessed PCH file?

running `g++ -std=c++1 path/to/header.hpp` works and creates a `.hpp.gch` file. It's binary so I don't really get what's in it.

**Comment by BenjaminW3** [comment]
`g++ -std=c++11 -E path/to/precompiled_header.hpp -o preprocessed.i`?

**Comment by ax3l** [comment]
That's something else. I am trying to do [*pre-compiled* headers](http://itscompiling.eu/2017/01/12/precompiled-headers-cpp-compilation/), `-E` is *pre-processed* only.

[AtomicOmpCritSec.hpp.gch.txt](https://github.com/ComputationalRadiationPhysics/alpaka/files/1228943/AtomicOmpCritSec.hpp.gch.txt)
[AtomicOmpCritSec.hpp.i.txt](https://github.com/ComputationalRadiationPhysics/alpaka/files/1228944/AtomicOmpCritSec.hpp.i.txt)

**Comment by BenjaminW3** [comment]
Everywhere I have seen precompiled headers being used, they were useful because multiple often used headers had been combined within a super-header which you then compile before the C++ files.
I wanted to see this preprocessed super-header.

**Comment by ax3l** [comment]
Yes, I am trying to create that one but I could not succeed due to the error above yet. Due to that error, the file is skipped and not included in that single super-header.

**Comment by BenjaminW3** [comment]
Do you still work on this or can we close this?

**Comment by ax3l** [comment]
Let us keep this open, we also just corrected a linking issue in PIConGPU which might have played into this.

If you don't mind, let us keep this in a "future" milestone - it's not urgent but a utterly useful feature if we might get it implemented.
